Update PULL_REQUEST_TEMPLATE.md

cyanglaz opened the pull request:
Fixes https://github.com/flutter/flutter/issues/80081

ditman commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -8,7 +8,7 @@
 - [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
 - [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
-- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
+- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that not like the flutter/flutter repo, we do use dart formatter in the flutter/plugins repo.)
```
```
[ ] I have formatted this PR with `dartfmt -w .`, read the [Flutter Style Guide] and followed the [C++, Objective-C, Java style guides].
```

Is there a plugins command that we can provide so users can quickly format all the source they might have touched in their PR? Something that guarantees that the `format` step will pass?

blasten replied:
does it make sense to say `dart run ./script/tool/lib/src/main.dart format` this those all of those.

cyanglaz replied:
done

blasten commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -8,7 +8,7 @@
 - [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
 - [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
-- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that not like the flutter/flutter repo, we do use dart formatter in the flutter/plugins repo.)
+- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that not like the flutter/flutter repo, we do use dart formatter in the flutter/plugins repo. Run `dart run ./script/tool/lib/src/main.dart format` under plugins/ to auto format all source code.)
```
> not like the -> unlike the

cyanglaz replied:
done

stuartmorgan replied:
This will only work if they have done a `pub get` there. I think we should put instructions in CONTRIBUTING.md, and then link directly to that part of it from here.

cyanglaz commented:
@stuartmorgan @ditman @blasten I updated the contribution guide to reflect the new tooling process we use, and updated the template to point user there.

blasten commented on CONTRIBUTING.md:
```diff
@@ -48,6 +48,14 @@ for any of the following plugins, we encourage you to submit it
 
+
+## Setting up tools
+
```
nit: extra line

cyanglaz replied:
done

blasten commented on CONTRIBUTING.md:
```diff
@@ -48,6 +48,14 @@ for any of the following plugins, we encourage you to submit it
+
+We use a range of tooling script to do varies of things on CI. (testing, formatting, etc.)
+Your are likely to use some of the tooling script locally in your contributing journey.
+See [plugin_tools](./script/tool/README.MD) for more details
```
```suggestion
See [plugin_tools](./script/tool/README.MD) for more details.
```

cyanglaz replied:
done

blasten commented on CONTRIBUTING.md:
```diff
@@ -159,12 +161,12 @@ To start working on a patch:
-pub global run flutter_plugin_tools format --plugins plugin_name
-pub global run flutter_plugin_tools analyze --plugins plugin_name
-pub global run flutter_plugin_tools test --plugins plugin_name
+$ cd script/tool & pub get
```
nit: no need for `$`

cyanglaz replied:
done

blasten commented on script/tool/README.md:
```diff
@@ -1,8 +1,60 @@
+
+``sh
+cd <path_to_plugins>/packages
+dart run ./script/tool/lib/src/main.dart publish-plugin --package $package
```
what is `$package`?  did you mean `<package>`?

cyanglaz replied:
yes, done

blasten commented on script/tool/README.md:
```diff
@@ -1,8 +1,60 @@
+dart run ./script/tool/lib/src/main.dart publish-plugin --package $package
+``
+
+By default the tool tries to push tags to the `upstream` remote, but that and
+some additional settings can be configured. Run `dart run ./script/tool/lib/src/main.dart publish-plugin --help` for more usage information.
```
"but that and some additional settings".

cyanglaz replied:
done

cyanglaz reviewed (commented):
@blasten Updated per your comments!

stuartmorgan commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -8,7 +8,7 @@
 - [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
 - [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
-- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
+- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, we do use dart formatter in the flutter/plugins repo. See [plugin_tool format](../script/tool/README.md#format-code))
```
Nit: \`dart format\`

stuartmorgan replied:
Nit: The structure would work better here if this said "Note that unlike the flutter/flutter repo, the flutter/plugins repo does use \`dart format\`."

stuartmorgan commented on CONTRIBUTING.md:
```diff
@@ -48,8 +45,14 @@ for any of the following plugins, we encourage you to submit it
+## Setting up tools
+
+We use a range of tooling script to do varies of things on CI. (testing, formatting, etc.)
+Your are likely to use some of the tooling script locally in your contributing journey.
```
How about just: "There are scripts for many common tasks (testing, formatting, etc.) that will likely be useful in preparing a PR.". It seems odd to frame it in terms of CI first when that's not the use case for most readers.

stuartmorgan commented on CONTRIBUTING.md:
````diff
@@ -159,12 +156,12 @@ To start working on a patch:
-pub global run flutter_plugin_tools test --plugins plugin_name
+ * Verify changes with [plugin_tools](./script/tool/README.MD).
+```sh
+cd script/tool && pub get
````
This is missing a `cd` back to the repo root.

stuartmorgan commented on CONTRIBUTING.md:
```diff
@@ -159,12 +156,12 @@ To start working on a patch:
+cd script/tool && pub get
+dart run ./script/tool/lib/src/main.dart format --plugins plugin_name
+pub run ./script/tool/lib/src/main.dart analyze --plugins plugin_name
+pub run ./script/tool/lib/src/main.dart test --plugins plugin_name
```
Shouldn't these all be `dart`?

stuartmorgan commented on CONTRIBUTING.md:
```diff
@@ -245,37 +242,4 @@ Releasing a package is a two-step process.
-There is a lot about this process that is still to be desired. Some top level
-items are being tracked in
-[flutter/flutter#27258](https://github.com/flutter/flutter/issues/27258).
+For convenience, a [publish-plugin](./script/tool/README.MD#publish-and-tag-release) tool script could be used to do both step at once.
```
Let's flip this section to say to use this script, and then give the manual steps as fallback. We should lead with the option we want people to use, and this is no longer experimental given that we use it all the time.

stuartmorgan commented on script/tool/README.md:
```diff
@@ -1,8 +1,60 @@
 # Flutter Plugin Tools
 
+Note: The commands in tools are designed to run under plugins/ or plugins/packages/.
```
s/under plugins/at the root of the repository/ (and similar for the second part); there's no guarantee that the local checkout is called `packages` even if it's the most likely scenario.

stuartmorgan commented on script/tool/README.md:
````diff
@@ -1,8 +1,60 @@
 
 ```sh
 dart pub get
 dart run lib/src/main.dart <args>
````
This command is misleading, given that as we've just said above this isn't the directory you should be in when running the tool.

stuartmorgan commented on script/tool/README.md:
````diff
@@ -1,8 +1,60 @@
+## Format Code
+
+```sh
+cd <path_to_plugins>/packages
````
Remove "packages" from all of these examples; it's not necessary to be there, and it makes the next line wrong.

stuartmorgan commented on script/tool/README.md:
````diff
@@ -1,8 +1,60 @@
+pub run ./script/tool/lib/src/main.dart analyze --plugins plugin_name
+```
+
+## Run dart unit tests
````
Nit: Dart

stuartmorgan commented on script/tool/README.md:
````diff
@@ -1,8 +1,60 @@
+```
+
+## Publish and tag release
+
````
Mention here that you should be at the commit of the change being published? I was doing this wrong for quite a while.

cyanglaz replied:
done. added a `git checkout` command in the sh sample

stuartmorgan commented on script/tool/README.md:
```diff
@@ -1,8 +1,61 @@
+
+``sh
+cd <path_to_plugins>
+git checkout <commit_hash_need_to_publish>
```
s/need_to/to/

stuartmorgan commented on script/tool/README.md:
```diff
@@ -1,8 +1,61 @@
+additional settings can be configured. Run `dart run ./script/tool/lib/src/main.dart publish-plugin --help` for more usage information.
+
+The tool wraps `pub publish` for pushing the package to pub, and then will
+automatically use git to try and create and push tags. It has some additional
```
Nit: "to try to create"

stuartmorgan commented on CONTRIBUTING.md:
```diff
@@ -48,8 +45,14 @@ for any of the following plugins, we encourage you to submit it
+## Setting up tools
+
+We use a range of tooling script to do varies of things on CI. (testing, formatting, etc.)
+There are scripts for many common tasks (testing, formatting, etc.) that will likely be useful in preparing a PR.
```
You replied Done to the comment about rewriting this, but it doesn't seem to have been applied here.

stuartmorgan reviewed (approved):
LGTM!

ditman reviewed (approved):
This is a great overhauling of the contributing process! Great stuff!

stuartmorgan commented on CONTRIBUTING.md:
```diff
@@ -48,8 +45,13 @@ for any of the following plugins, we encourage you to submit it
+## Setting up tools
+
+There are scripts for many common tasks (testing, formatting, etc.) that will likely be useful in preparing a PR.
+See [plugin_tools](./script/tool/README.MD) for more details.
```
Can you scrub this for all instances of `.MD` and replace them with `.md`? I just got a 404 from one of these links due to the incorrect capitalization.

cyanglaz replied:
Good catch! Thanks! Fix is here: https://github.com/flutter/plugins/pull/3815
